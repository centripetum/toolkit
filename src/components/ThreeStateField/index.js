import * as React from 'react'

import { identity } from 'ramda'

const { useState } = React

function ThreeStateField ({
  trueField,
  falseField,
  defaultValue = null,
  validate = identity
}) {
  var [value, setValue] = useState(defaultValue)
  const [valid, setValid] = useState(true)

  const handleChange = ({ target = {} }) => {
    if (target.id === 'trueBox') {
      value = true
      if (document.getElementById('falseBox').checked) {
        document.getElementById('falseBox').checked = false
        value = true
      } else if (document.getElementById('trueBox').checked) {
        value = true
      } else {
        value = null
      }
    } else if (target.id === 'falseBox') {
      if (document.getElementById('trueBox').checked) {
        document.getElementById('trueBox').checked = false
        value = false
      } else if (document.getElementById('falseBox').checked) {
        value = false
      } else {
        value = null
      }
    }
    setValue(value)
    setValid(validate(value))
  }

  return (
    <div>
      {trueField}
      <input
        id='trueBox'
        type='checkbox'
        value={value}
        onChange={handleChange}
      />
      <br />
      {falseField}
      <input
        id='falseBox'
        type='checkbox'
        value={value}
        onChange={handleChange}
      />
    </div>
  )
}

export default ThreeStateField
