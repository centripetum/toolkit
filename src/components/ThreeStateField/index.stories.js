import * as React from 'react'
import { storiesOf } from '@storybook/react'

import ThreeStateField from './'

const field1 = 'Yes'
const field2 = 'No'

storiesOf('components/ThreeStateField', module).add('Basic component', () => (
  <ThreeStateField field1={field1} field2={field2} />
))
