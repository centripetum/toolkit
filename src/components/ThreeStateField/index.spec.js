import * as React from 'react'
import { render } from 'react-testing-library'

import ThreeStateField from './'

describe('component: ThreeStateField', () => {
  it('renders a three state field', () => {
    const { container } = render(
      <ThreeStateField trueField={'True'} falseField={'False'} />
    )
    expect(container).toMatchSnapshot()
  })
})
