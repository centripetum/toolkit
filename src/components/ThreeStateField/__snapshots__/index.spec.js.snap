// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: ThreeStateField renders a three state field 1`] = `
<div>
  <div>
    True
    <input
      id="trueBox"
      type="checkbox"
      value=""
    />
    <br />
    False
    <input
      id="falseBox"
      type="checkbox"
      value=""
    />
  </div>
</div>
`;
